File: src/modules/system/versions.ts
import { Elysia, t } from 'elysia'
import { VersionManager, type VersionInfo } from './version-manager'

// Response models
const VersionInfoModel = t.Object({
  version: t.String(),
  path: t.String(),
  installed: t.Boolean(),
  isLatest: t.Boolean(),
  size: t.Optional(t.Number()),
  installedAt: t.Optional(t.Date())
})

const VersionListModel = t.Array(VersionInfoModel)

const ErrorModel = t.Object({
  error: t.String(),
  success: t.Literal(false)
})

const SuccessModel = t.Object({
  success: t.Literal(true),
  message: t.String()
})

const InstallRequestModel = t.Object({
  version: t.String()
})

const VersionSizesModel = t.Record(t.String(), t.Number())

export const versionsModule = new Elysia({ prefix: '/versions' })
  // Get all installed versions
  .get('/installed', async () => {
    try {
      const versions = await VersionManager.getInstalledVersions()
      return versions
    } catch (error) {
      console.error('Failed to get installed versions:', error)
      return { error: 'Failed to get installed versions', success: false }
    }
  }, {
    response: {
      200: VersionListModel,
      500: ErrorModel
    }
  })

  // Get available versions from GitHub
  .get('/available', async ({ query }) => {
    try {
      const limit = query.limit ? parseInt(query.limit as string) : 10
      const versions = await VersionManager.getAvailableVersions(limit)
      return versions
    } catch (error) {
      console.error('Failed to get available versions:', error)
      return { error: 'Failed to get available versions', success: false }
    }
  }, {
    query: t.Object({
      limit: t.Optional(t.String())
    }),
    response: {
      200: VersionListModel,
      500: ErrorModel
    }
  })

  // Install a specific version
  .post('/install', async ({ body, set }) => {
    try {
      await VersionManager.installVersion(body.version)
      return { 
        success: true, 
        message: `Successfully installed GOWA version ${body.version}` 
      }
    } catch (error) {
      set.status = 500
      return { 
        error: error instanceof Error ? error.message : 'Failed to install version', 
        success: false 
      }
    }
  }, {
    body: InstallRequestModel,
    response: {
      200: SuccessModel,
      500: ErrorModel
    }
  })

  // Remove a specific version
  .delete('/:version', async ({ params: { version }, set }) => {
    try {
      if (version === 'latest') {
        set.status = 400
        return { 
          error: 'Cannot remove the latest version alias', 
          success: false 
        }
      }

      await VersionManager.removeVersion(version)
      return { 
        success: true, 
        message: `Successfully removed GOWA version ${version}` 
      }
    } catch (error) {
      set.status = 500
      return { 
        error: error instanceof Error ? error.message : 'Failed to remove version', 
        success: false 
      }
    }
  }, {
    response: {
      200: SuccessModel,
      400: ErrorModel,
      500: ErrorModel
    }
  })

  // Check if a version is available/installed
  .get('/:version/available', async ({ params: { version } }) => {
    try {
      const available = await VersionManager.isVersionAvailable(version)
      return { 
        version, 
        available,
        path: VersionManager.getVersionBinaryPath(version)
      }
    } catch (error) {
      return { 
        error: error instanceof Error ? error.message : 'Failed to check version availability', 
        success: false 
      }
    }
  }, {
    response: {
      200: t.Object({
        version: t.String(),
        available: t.Boolean(),
        path: t.String()
      }),
      500: ErrorModel
    }
  })

  // Get disk usage for all versions
  .get('/usage', async () => {
    try {
      const sizes = await VersionManager.getVersionsSize()
      return sizes
    } catch (error) {
      return { 
        error: error instanceof Error ? error.message : 'Failed to get version sizes', 
        success: false 
      }
    }
  }, {
    response: {
      200: VersionSizesModel,
      500: ErrorModel
    }
  })

  // Cleanup old versions
  .post('/cleanup', async ({ body, set }) => {
    try {
      const keepCount = body?.keepCount ?? 3
      const removed = await VersionManager.cleanup(keepCount)
      return { 
        success: true, 
        message: `Cleaned up ${removed.length} old versions: ${removed.join(', ')}`,
        removed
      }
    } catch (error) {
      set.status = 500
      return { 
        error: error instanceof Error ? error.message : 'Failed to cleanup versions', 
        success: false 
      }
    }
  }, {
    body: t.Optional(t.Object({
      keepCount: t.Optional(t.Number({ minimum: 1, maximum: 10 }))
    })),
    response: {
      200: t.Object({
        success: t.Literal(true),
        message: t.String(),
        removed: t.Array(t.String())
      }),
      500: ErrorModel
    }
  })

File: internal/httpapi/versions.go
package httpapi

import (
	"context"
	"errors"
	"io"
	"net/http"
	"strconv"
	"strings"
	"time"

	"github.com/fadlee/gowa-manager/internal/versions"
)

var (
	ErrVersionConflict = errors.New("version conflict")
	ErrVersionInvalid  = errors.New("invalid version")
)

type VersionService interface {
	GetInstalledVersions() ([]versions.VersionInfo, error)
	GetAvailableVersions(context.Context, int) ([]versions.VersionInfo, error)
	IsVersionAvailable(context.Context, string) (bool, error)
	GetVersionsSize() (map[string]int64, error)
	RemoveVersion(string) error
	Cleanup(int) ([]string, error)
}

type VersionInstaller interface {
	Install(context.Context, string) (versions.InstallResult, error)
}

type VersionInfo struct {
	Version     string    `json:"version"`
	Path        string    `json:"path"`
	Installed   bool      `json:"installed"`
	IsLatest    bool      `json:"isLatest"`
	Size        int64     `json:"size"`
	InstalledAt time.Time `json:"installedAt"`
}

func registerVersionRoutes(mux *http.ServeMux, deps Dependencies) {
	h := &versionHandler{service: deps.Versions, installer: deps.VersionInstaller}
	mux.HandleFunc("/api/system/versions/installed", h.installed)
	mux.HandleFunc("/api/system/versions/available", h.available)
	mux.HandleFunc("/api/system/versions/install", h.install)
	mux.HandleFunc("/api/system/versions/usage", h.usage)
	mux.HandleFunc("/api/system/versions/cleanup", h.cleanup)
	mux.HandleFunc("/api/system/versions/", h.routes)
}

type versionHandler struct {
	service   VersionService
	installer VersionInstaller
}

func (h *versionHandler) installed(w http.ResponseWriter, r *http.Request) {
	if r.Method != http.MethodGet {
		w.WriteHeader(http.StatusMethodNotAllowed)
		return
	}
	items, err := h.service.GetInstalledVersions()
	if err != nil {
		writeHTTPError(w, http.StatusInternalServerError, err)
		return
	}
	writeJSON(w, http.StatusOK, toVersionInfoResponses(items))
}

func (h *versionHandler) available(w http.ResponseWriter, r *http.Request) {
	if r.Method != http.MethodGet {
		w.WriteHeader(http.StatusMethodNotAllowed)
		return
	}
	limit := 10
	if raw := r.URL.Query().Get("limit"); raw != "" {
		parsed, err := strconv.Atoi(raw)
		if err != nil || parsed < 1 {
			writeValidation(w, "Invalid limit")
			return
		}
		limit = parsed
	}
	items, err := h.service.GetAvailableVersions(r.Context(), limit)
	if err != nil {
		writeHTTPError(w, http.StatusInternalServerError, err)
		return
	}
	writeJSON(w, http.StatusOK, toVersionInfoResponses(items))
}

func toVersionInfoResponses(items []versions.VersionInfo) []VersionInfo {
	out := make([]VersionInfo, 0, len(items))
	for _, item := range items {
		out = append(out, VersionInfo{Version: item.Version, Path: item.Path, Installed: item.Installed, IsLatest: item.IsLatest, Size: item.Size, InstalledAt: item.InstalledAt})
	}
	return out
}

func (h *versionHandler) install(w http.ResponseWriter, r *http.Request) {
	if r.Method != http.MethodPost {
		w.WriteHeader(http.StatusMethodNotAllowed)
		return
	}
	var body struct {
		Version string `json:"version"`
	}
	if err := decodeJSON(r, &body); err != nil {
		writeValidation(w, "Invalid JSON")
		return
	}
	if strings.TrimSpace(body.Version) == "" {
		writeValidation(w, "version is required")
		return
	}
	result, err := h.installer.Install(r.Context(), body.Version)
	if err != nil {
		h.writeVersionError(w, err)
		return
	}
	writeJSON(w, http.StatusOK, struct {
		Success          bool   `json:"success"`
		Message          string `json:"message"`
		Version          string `json:"version"`
		Path             string `json:"path"`
		SHA256           string `json:"sha256"`
		Size             int64  `json:"size"`
		AlreadyInstalled bool   `json:"already_installed"`
	}{Success: true, Message: "Version installed successfully", Version: result.Version, Path: result.Path, SHA256: result.SHA256, Size: result.Size, AlreadyInstalled: result.AlreadyInstalled})
}

func (h *versionHandler) usage(w http.ResponseWriter, r *http.Request) {
	if r.Method != http.MethodGet {
		w.WriteHeader(http.StatusMethodNotAllowed)
		return
	}
	sizes, err := h.service.GetVersionsSize()
	if err != nil {
		writeHTTPError(w, http.StatusInternalServerError, err)
		return
	}
	writeJSON(w, http.StatusOK, struct {
		Versions map[string]int64 `json:"versions"`
	}{Versions: sizes})
}

func (h *versionHandler) cleanup(w http.ResponseWriter, r *http.Request) {
	if r.Method != http.MethodPost {
		w.WriteHeader(http.StatusMethodNotAllowed)
		return
	}
	keepCount := 3
	if r.Body != nil {
		var body struct {
			KeepCount *int `json:"keepCount"`
		}
		if err := decodeJSON(r, &body); err != nil && !errors.Is(err, io.EOF) {
			writeValidation(w, "Invalid JSON")
			return
		}
		if body.KeepCount != nil {
			keepCount = *body.KeepCount
		}
	}
	if keepCount < 1 {
		writeValidation(w, "keepCount must be at least 1")
		return
	}
	removed, err := h.service.Cleanup(keepCount)
	if err != nil {
		h.writeVersionError(w, err)
		return
	}
	writeJSON(w, http.StatusOK, struct {
		Success bool     `json:"success"`
		Message string   `json:"message"`
		Removed []string `json:"removed"`
	}{Success: true, Message: "Version cleanup completed successfully", Removed: removed})
}

func (h *versionHandler) routes(w http.ResponseWriter, r *http.Request) {
	path := strings.TrimPrefix(r.URL.Path, "/api/system/versions/")
	if strings.HasSuffix(path, "/available") {
		version := strings.TrimSuffix(path, "/available")
		h.versionAvailable(w, r, version)
		return
	}
	if strings.Contains(path, "/") || path == "" {
		writeJSON(w, http.StatusNotFound, map[string]any{"success": false, "error": "Not found"})
		return
	}
	if r.Method != http.MethodDelete {
		w.WriteHeader(http.StatusMethodNotAllowed)
		return
	}
	if path == "latest" {
		writeValidation(w, "cannot remove latest version")
		return
	}
	if err := h.service.RemoveVersion(path); err != nil {
		h.writeVersionError(w, err)
		return
	}
	writeJSON(w, http.StatusOK, map[string]any{"success": true, "message": "Version removed successfully"})
}

func (h *versionHandler) versionAvailable(w http.ResponseWriter, r *http.Request, version string) {
	if r.Method != http.MethodGet {
		w.WriteHeader(http.StatusMethodNotAllowed)
		return
	}
	if version == "" || strings.Contains(version, "/") {
		writeValidation(w, "Invalid version")
		return
	}
	available, err := h.service.IsVersionAvailable(r.Context(), version)
	if err != nil {
		h.writeVersionError(w, err)
		return
	}
	writeJSON(w, http.StatusOK, struct {
		Version   string `json:"version"`
		Available bool   `json:"available"`
	}{Version: version, Available: available})
}

func (h *versionHandler) writeVersionError(w http.ResponseWriter, err error) {
	if errors.Is(err, ErrVersionConflict) || errors.Is(err, ErrVersionInvalid) || strings.Contains(err.Error(), "cannot remove active version") || strings.Contains(err.Error(), "invalid version") {
		writeHTTPError(w, http.StatusBadRequest, err)
		return
	}
	writeHTTPError(w, http.StatusInternalServerError, err)
}
